### PR TITLE
Rearranging tabs: Feature announcement banner

### PR DESCRIPTION
**Pull request by 0nko**
Task/Issue URL: https://app.asana.com/0/1207418217763355/1207737066387223/f

### Description

This PR adds a feature announcement banner to the tab switcher screen. It will be visible at the top, no matter how many tabs or where the scroll position is. The following logic dictates when the banner is displayed:

- Only existing users will see the banner
- Unless it is explicitly dismissed, it will be shown 3 times
- Dismissing the banner will permanently hide it
- Performing the drag & drop action will permanently hide the banner
- The banner will not be displayed unless there are at least 2 tabs

Note: The announcement copy is hardcoded while it's in review. After it's approved, a new PR will be opened with the translations.

### Steps to test this PR

_New users_
- [x] Uninstall & reinstall the app
- [x] Delete the `DuckDuckGo` folder from `Downloads`, if exists
- [x] Start the app and go to the tab switcher
- [x] Notice the banner is not visible
- [x] Add at least 2 tabs
- [x] Notice the banner is still not visible

_Reinstalls_
- [x] Start the app
- [x] Close it and uninstall it
- [x] Reinstall the app and start it
- [x] Go to the tab switcher
- [x] Add at least 2 tabs
- [x] Notice the banner *is* visible

_Existing users (displaying 3 times)_
- [x] Make sure the app has been previously installed before running this test (if not, reinstall it again)
- [x] Add at least 2 tabs (long press on the tabs toolbar button)
- [x] Go to the tab switcher
- [x] Notice the announcement banner is displayed (the 1st time)
- [x] Go back to the browser
- [x] Open the tab switcher again
- [x] Notice the announcement banner is displayed (the 2nd time)
- [x] Go back to the browser
- [x] Open the tab switcher again
- [x] Notice the announcement banner is displayed (the 3rd time)
- [x] Go back to the browser
- [x] Open the tab switcher again
- [x] Notice the announcement banner is not displayed anymore

_Existing users (dismissing the banner)_
- [x] Make sure the app has been previously installed before running this test (if not, reinstall it again)
- [x] Add at least 2 tabs (long press on the tabs toolbar button)
- [x] Go to the tab switcher
- [x] Notice the announcement banner is displayed
- [x] Tap on the Close button on the banner
- [x] Notice the banner is dismissed
- [x] Go back to the browser
- [x] Open the tab switcher again
- [x] Notice the announcement banner is not displayed anymore

_Existing users (no tabs)_
- [x] Make sure the app has been previously installed before running this test (if not, reinstall it again)
- [x] Start the app and go to the tab switcher
- [x] Notice that the banner is not visible
- [x] Add at least 2 tabs
- [x] Notice the banner appears
- [x] Remove the tabs
- [x] Notice the banner does not disappear if there are fewer than 2 tabs
- [x] Leave the tab manager and come back
- [x] Notice the banner is not visible
- [x] Start adding tabs and notice the banner is displayed when there 2 or more tabs

_Existing users (performing the action)_
- [x] Make sure the app has been previously installed before running this test (if not, reinstall it again)
- [x] Add at least 2 tabs (long press on the tabs toolbar button)
- [x] Go to the tab switcher
- [x] Notice that the banner is visible
- [x] Long-tap on a tab and drag it somewhere else
- [x] Notice the banner disappears

### UI
| Light  | Dark |
| ------ | ----- |
||

### CODE REVIEW

**Comment by 0nko** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4743?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4743** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4743?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4734** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4734?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4719** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4719?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feature/ondrej/rearranging-tabs`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @0nko and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherActivity.kt [hunk 1]
```diff
@@ -167,6 +183,20 @@ class TabSwitcherActivity : DuckDuckGoActivity(), TabSwitcherListener, Coroutine
...
             }
         }
+
+        lifecycleScope.launch {
+            viewModel.isFeatureAnnouncementVisible
+                .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
+                .collect { isVisible ->
+                    if (isVisible) {
+                        viewModel.onTabFeatureAnnouncementDisplayed()
+                        announcement.visibility = View.VISIBLE
```
Comment (anikiki) [inline app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherActivity.kt:193]: You can use `announcement.show()` / `announcement.gone()`.

File: app/src/main/res/layout/content_tab_switcher.xml [hunk 2]
```diff
@@ -13,22 +13,69 @@
...
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="horizontal">
+
+            <com.duckduckgo.common.ui.view.text.DaxTextView
+                android:layout_width="0dp"
+                android:layout_height="wrap_content"
+                android:layout_weight="1"
+                android:layout_margin="16dp"
```
Comment (anikiki) [inline app/src/main/res/layout/content_tab_switcher.xml:45]: You can use existing `dimens` for this layout. 
See https://github.com/duckduckgo/Android/blob/develop/common/common-ui/src/main/res/values/design-system-dimensions.xml

File: browser-api/src/main/java/com/duckduckgo/app/tabs/model/TabSwitcherData.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+package com.duckduckgo.app.tabs.model
+
+data class TabSwitcherData(
+    val userState: UserState,
+    val wasAnnouncementDismissed: Boolean,
+    val announcementDisplayCount: Int,
+) {
+    enum class UserState {
+        NEW,
+        RETURNING,
```
Comment (anikiki) [inline browser-api/src/main/java/com/duckduckgo/app/tabs/model/TabSwitcherData.kt:26]: Maybe rename this to `EXISTING`.

File: app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherActivity.kt (hunk 1 above)
Comment (0nko) [reply app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherActivity.kt]: Done in [6054d81](https://github.com/duckduckgo/Android/pull/4743/commits/6054d81879b1a47d62ea0bb465ff877e474068a6).

File: browser-api/src/main/java/com/duckduckgo/app/tabs/model/TabSwitcherData.kt (hunk 3 above)
Comment (0nko) [reply browser-api/src/main/java/com/duckduckgo/app/tabs/model/TabSwitcherData.kt]: Done in [81e783b](https://github.com/duckduckgo/Android/pull/4743/commits/81e783b658c4c21035e19a9d4f742c5e60596e1e).

**Review by anikiki — APPROVED** [review]
LGTM :+1:

File: app/src/main/java/com/duckduckgo/app/launch/LaunchViewModel.kt [hunk 4]
```diff
@@ -44,6 +46,7 @@ class LaunchViewModel @Inject constructor(
 
     suspend fun determineViewToShow() {
         waitForReferrerData()
+        tabDataRepository.setIsUserNew(userStageStore.isNewUser())
```
Comment (malmstein) [inline app/src/main/java/com/duckduckgo/app/launch/LaunchViewModel.kt:49]: sorry I’m late to the party, but this doesn’t belong here. What does the LaunchViewModel have to do with Tabs? There are better ways to make a default value based on the stage.

**Review by malmstein — COMMENTED** [review]
You’re leaking implementation details of tabs to a `ViewModel` that has nothing to do with it.

**Review by malmstein — CHANGES_REQUESTED** [review]
Clean up the `LaunchViewModel` so there implementation details of Tabs are not leaked.

**Comment by 0nko** [comment]
@malmstein, @anikiki I’ve refactored the way we query for new users in [a023d94](https://github.com/duckduckgo/Android/pull/4743/commits/a023d94bb7ae913d418ca31e97ac19bcff532423). The logic is a little bit different but for our purposes it should have the same result. 

To account for returning user, there’s now an additional condition in the tab manager ([27d86fb](https://github.com/duckduckgo/Android/pull/4743/commits/27d86fbfaafd1a4149b9b71d4cc8b9c796eaf051)).

Just to recapitulate what the definition of a new user is here — a new user is someone who's never installed the app (so not a returning user).

**Review by anikiki — APPROVED** [review]
LGTM :+1:

**Pull request merged**